src/core/data/transform.cc

```diff
@@ -239,6 +246,11 @@ void Project::print(std::ostream& out) const
 
 int32_t Project::target_ndim(int32_t source_ndim) const { return source_ndim + 1; }
 
+void Project::return_promoted_dims(std::vector<int32_t>& dims) const
+{
+  dims.push_back((-dim_ - 1));
```

Comment:
If you process the parent before the child, then I don't think you need to leave this "sentinel" value, you can just remove the projected-out dimension from the `promoted_dims` that the parent returned (if it's there), and shift all larger dims down by one.